zerver/lib/actions.py

```diff
@@ -1632,18 +1634,18 @@ def bulk_insert_ums(ums: List[UserMessageLite]) -> None:
     if not ums:
         return
 
-    vals = ','.join([
-        '(%d, %d, %d)' % (um.user_profile_id, um.message_id, um.flags)
+    vals = [
+        (um.user_profile_id, um.message_id, um.flags)
         for um in ums
-    ])
-    query = '''
+    ]
+    query = SQL('''
         INSERT into
             zerver_usermessage (user_profile_id, message_id, flags)
-        VALUES
-    ''' + vals
+        VALUES %s
+    ''')
 
     with connection.cursor() as cursor:
-        cursor.execute(query)
+        execute_values(cursor.cursor, query, vals)
```

Comment:
How does this work?  I'm a bit concerned that this will result in the query not being counted in our logging (`zerver/lib/db.py`)

Reply:
`cursor` is a `django.db.backends.utils.CursorWrapper`; `cursor.cursor` is a `zerver.lib.db.TimeTrackingCursor`. So I think we still get the logging.

The problem being solved here is that `execute_values` contains an `isinstance` check that is satisfied by `TimeTrackingCursor`, which is a subclass of `psycopg2.extensions.cursor`, but not by `CursorWrapper`, which is not. It’s all very annoying.